perf/collector_libpfm.go

```diff
@@ -152,22 +181,71 @@ func (c *collector) setup() error {
 	return nil
 }
 
-func (c *collector) setupRawNonGrouped(event *CustomEvent, cgroup int) error {
-	klog.V(5).Infof("Setting up non-grouped raw perf event %#v", event)
+func (c *collector) setupRawEvent(event *CustomEvent, cgroup int, index int, leaderFileDescriptors *[]int, leader bool) error {
+	klog.V(5).Infof("Setting up grouped raw perf event %#v", event)
+
 	config := createPerfEventAttr(*event)
-	err := c.registerEvent(config, string(event.Name), cgroup)
+	setGroupAttributes(config, leader)
+	klog.V(5).Infof("perf_event_attr struct prepared: %#v", config)
+
+	err := c.registerEvent(config, string(event.Name), cgroup, index, leaderFileDescriptors, leader)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (c *collector) setupEvent(name string, cgroup int, index int, leaderFileDescriptors *[]int, leader bool) error {
+	if !isLibpfmInitialized {
+		return fmt.Errorf("libpfm4 is not initialized, cannot proceed with setting perf events up")
+	}
+
+	klog.V(5).Infof("Setting up grouped perf event %s", name)
+
+	perfEventAttrMemory := C.malloc(C.ulong(unsafe.Sizeof(unix.PerfEventAttr{})))
+	defer C.free(perfEventAttrMemory)
+	err := readPerfEventAttr(name, perfEventAttrMemory)
 	if err != nil {
 		return err
 	}
+	perfEventAttr := (*unix.PerfEventAttr)(perfEventAttrMemory)
+
+	setGroupAttributes(perfEventAttr, leader)
+
+	klog.V(5).Infof("perf_event_attr: %#v", perfEventAttr)
+	return c.registerEvent(perfEventAttr, name, cgroup, index, leaderFileDescriptors, leader)
+}
 
+func readPerfEventAttr(name string, perfEventAttrMemory unsafe.Pointer) error {
+	event := pfmPerfEncodeArgT{}
+	fstr := C.CString("")
+	event.fstr = unsafe.Pointer(fstr)
+	event.attr = perfEventAttrMemory
+	event.size = C.ulong(unsafe.Sizeof(event))
+	cSafeName := C.CString(name)
+	pErr := C.pfm_get_os_event_encoding(cSafeName, C.PFM_PLM0|C.PFM_PLM3, C.PFM_OS_PERF_EVENT, unsafe.Pointer(&event))
+	if pErr != C.PFM_SUCCESS {
+		return fmt.Errorf("unable to transform event name %s to perf_event_attr: %d", name, int(pErr))
+	}
 	return nil
 }
 
-func (c *collector) registerEvent(config *unix.PerfEventAttr, name string, pid int) error {
-	var cpu int
-	for cpu = 0; cpu < c.numCores; cpu++ {
-		groupFd, flags := -1, unix.PERF_FLAG_FD_CLOEXEC|unix.PERF_FLAG_PID_CGROUP
-		fd, err := unix.PerfEventOpen(config, pid, cpu, groupFd, flags)
+func (c *collector) registerEvent(config *unix.PerfEventAttr, name string, cgroup int, groupIndex int, leaderFileDescriptors *[]int, leader bool) error {
+	for cpu := 0; cpu < c.numCores; cpu++ {
+		var groupFD, pid, flags int
+		if leader {
+			// Leader should have this parameter set for -1.
+			groupFD = -1
+			pid = cgroup
+			flags = unix.PERF_FLAG_FD_CLOEXEC | unix.PERF_FLAG_PID_CGROUP
+		} else {
+			// If not leader, groupFd should be set to leaders perf file fd.
+			groupFD = (*leaderFileDescriptors)[cpu]
```

Comment:
You make a dangerous assumption: for first iteration of `for cpu := 0; cpu < c.numCores; cpu++` loop must `leader` must be `true` - otherwise you will try retrieve non-initialized item from `leaderFileDescriptors`. This error might be subtle as `leaderFileDescriptors` is already initialized using `make` and statement will not panic for `cpu < c.numCores`
Perhaps:
* remove: 
```go
if leader {
    config.Bits |= unix.PerfBitDisabled
}
```` 
from `setGroupAttributes`
* do not set `groupLeader` in `setup`
* decide who the leader is in `registerEvent` (and be careful: on ARM you can disable CPU 0) and set `config.Bits` accordingly.

Such approach would allow us to make a decision where we need to know a leader.
CC: @dashpole @katarzyna-z